src/common/storage/backends/s3/Client.js

```diff
@@ -0,0 +1,178 @@
+/* globals define */
+define([
+    '../StorageClient'
+], function (
+    StorageClient,
+) {
+    const S3Storage = function (id, name, logger, config = {}) {
+        StorageClient.apply(this, arguments);
+        this.bucketName = config.bucketName || 'deepforge';
+        this.endpoint = config.endpoint || 'http://localhost:80';
+        this.config = this.createS3Config(config);
+        this.s3Client = null;
+        this.ready = this.initialize();
+    };
+
+    S3Storage.prototype = Object.create(StorageClient.prototype);
+
+
+    S3Storage.prototype.initialize = async function () {
+        if (require.isBrowser) {
+            await new Promise((resolve, reject) => {
+                require(['aws-sdk-min'], () => {
+                    this.AWS = window.AWS;
+                    resolve();
+                }, reject);
+            });
+        } else {
+            this.AWS = require.nodeRequire('aws-sdk');
+        }
+    };
+
+    S3Storage.prototype.initializeS3Client = function (AWS, config) {
+        const s3Client = new AWS.S3(config);
+        promisifyMethod(s3Client, 'createBucket');
+        promisifyMethod(s3Client, 'getObject');
+        promisifyMethod(s3Client, 'putObject');
+        promisifyMethod(s3Client, 'deleteObject');
+        promisifyMethod(s3Client, 'headObject');
+        promisifyMethod(s3Client, 'listObjectsV2');
+        return s3Client;
+    };
+
+    S3Storage.prototype.getS3Client = async function (config) {
+        await this.ready;
+        if(!config){
+            if(!this.s3Client){
+                this.s3Client = this.initializeS3Client(this.AWS, this.config);
+            }
+            return this.s3Client;
+        } else {
+            config = this.createS3Config(config);
+            return this.initializeS3Client(this.AWS, config);
+        }
+    };
+
+    S3Storage.prototype.createS3Config = function (config) {
+        return {
+            endpoint: config.endpoint || 'http://localhost:80',
```

Comment:
Do we want localhost to be the default? Generally, we don't want users to use localhost whenever possible because this will not work on most compute backends

Reply:
We could choose `https://s3.amazonaws.com` as default? Or `https://play.min.io`?

Reply:
The default now is `https://s3.amazonaws.com`